File: farebot-card-cepas/src/main/java/com/codebutler/farebot/card/cepas/CEPASCard.java
/*
 * CEPASCard.java
 *
 * This file is part of FareBot.
 * Learn more at: https://codebutler.github.io/farebot/
 *
 * Copyright (C) 2011-2012, 2014, 2016 Eric Butler <eric@codebutler.com>
 * Copyright (C) 2011 Sean Cross <sean@chumby.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.card.cepas;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.codebutler.farebot.card.Card;
import com.codebutler.farebot.card.CardType;
import com.codebutler.farebot.base.ui.FareBotUiTree;
import com.codebutler.farebot.base.util.ByteArray;
import com.google.auto.value.AutoValue;

import java.text.DateFormat;
import java.text.NumberFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

@AutoValue
public abstract class CEPASCard extends Card {

    @NonNull
    public static CEPASCard create(
            @NonNull ByteArray tagId,
            @NonNull Date scannedAt,
            @NonNull List<CEPASPurse> purses,
            @NonNull List<CEPASHistory> histories) {
        return new AutoValue_CEPASCard(tagId, scannedAt, purses, histories);
    }

    @NonNull
    @Override
    public CardType getCardType() {
        return CardType.CEPAS;
    }

    @NonNull
    public abstract List<CEPASPurse> getPurses();

    @NonNull
    public abstract List<CEPASHistory> getHistories();

    @Nullable
    public CEPASPurse getPurse(int purse) {
        return getPurses().get(purse);
    }

    @Nullable
    public CEPASHistory getHistory(int purse) {
        return getHistories().get(purse);
    }

    @NonNull
    @Override
    public FareBotUiTree getAdvancedUi(Context context) {
        FareBotUiTree.Builder cardUiBuilder = FareBotUiTree.builder(context);

        FareBotUiTree.Item.Builder pursesUiBuilder = cardUiBuilder.item().title("Purses");
        for (CEPASPurse purse : getPurses()) {
            FareBotUiTree.Item.Builder purseUiBuilder = pursesUiBuilder.item()
                    .title(String.format("Purse ID %s", purse.getId()));
            purseUiBuilder.item().title("CEPAS Version").value(purse.getCepasVersion());
            purseUiBuilder.item().title("Purse Status").value(purse.getPurseStatus());
            purseUiBuilder.item().title("Purse Balance")
                    .value(NumberFormat.getCurrencyInstance(Locale.US).format(purse.getPurseBalance() / 100.0));
            purseUiBuilder.item().title("Purse Creation Date")
                    .value(DateFormat.getDateInstance(DateFormat.LONG).format(purse.getPurseCreationDate() * 1000L));
            purseUiBuilder.item().title("Purse Expiry Date")
                    .value(DateFormat.getDateInstance(DateFormat.LONG).format(purse.getPurseExpiryDate() * 1000L));
            purseUiBuilder.item().title("Autoload Amount").value(purse.getAutoLoadAmount());
            purseUiBuilder.item().title("CAN").value(purse.getCAN());
            purseUiBuilder.item().title("CSN").value(purse.getCSN());

            FareBotUiTree.Item.Builder transactionUiBuilder
                    = cardUiBuilder.item().title("Last Transaction Information");
            transactionUiBuilder.item().title("TRP").value(purse.getLastTransactionTRP());
            transactionUiBuilder.item().title("Credit TRP").value(purse.getLastCreditTransactionTRP());
            transactionUiBuilder.item().title("Credit Header").value(purse.getLastCreditTransactionHeader());
            transactionUiBuilder.item().title("Debit Options").value(purse.getLastTransactionDebitOptionsByte());

            FareBotUiTree.Item.Builder otherUiBuilder = cardUiBuilder.item().title("Other Purse Information");
            otherUiBuilder.item().title("Logfile Record Count").value(purse.getLogfileRecordCount());
            otherUiBuilder.item().title("Issuer Data Length").value(purse.getIssuerDataLength());
            otherUiBuilder.item().title("Issuer-specific Data").value(purse.getIssuerSpecificData());
        }

        return cardUiBuilder.build();
    }
}

File: farebot-card-cepas/src/commonMain/kotlin/com/codebutler/farebot/card/cepas/CEPASCard.kt
/*
 * CEPASCard.kt
 *
 * This file is part of FareBot.
 * Learn more at: https://codebutler.github.io/farebot/
 *
 * Copyright (C) 2011-2012, 2014, 2016 Eric Butler <eric@codebutler.com>
 * Copyright (C) 2011 Sean Cross <sean@chumby.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.codebutler.farebot.card.cepas

import com.codebutler.farebot.card.Card
import com.codebutler.farebot.card.CardType
import com.codebutler.farebot.base.ui.FareBotUiTree
import com.codebutler.farebot.base.util.DateFormatStyle
import com.codebutler.farebot.base.util.StringResource
import com.codebutler.farebot.base.util.formatDate
import kotlin.time.Instant
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import com.codebutler.farebot.base.util.CurrencyFormatter

@Serializable
data class CEPASCard(
    @Contextual override val tagId: ByteArray,
    override val scannedAt: Instant,
    val purses: List<CEPASPurse>,
    val histories: List<CEPASHistory>
) : Card() {

    override val cardType: CardType = CardType.CEPAS

    fun getPurse(purse: Int): CEPASPurse? = purses[purse]

    fun getHistory(purse: Int): CEPASHistory? = histories[purse]

    override fun getAdvancedUi(stringResource: StringResource): FareBotUiTree {
        val cardUiBuilder = FareBotUiTree.builder(stringResource)

        val pursesUiBuilder = cardUiBuilder.item().title("Purses")
        for (purse in purses) {
            val purseUiBuilder = pursesUiBuilder.item()
                .title("Purse ID ${purse.id}")
            purseUiBuilder.item().title("CEPAS Version").value(purse.cepasVersion)
            purseUiBuilder.item().title("Purse Status").value(purse.purseStatus)
            purseUiBuilder.item().title("Purse Balance")
                .value(CurrencyFormatter.formatValue(purse.purseBalance / 100.0, "SGD"))
            purseUiBuilder.item().title("Purse Creation Date")
                .value(formatDate(Instant.fromEpochMilliseconds(purse.purseCreationDate * 1000L), DateFormatStyle.LONG))
            purseUiBuilder.item().title("Purse Expiry Date")
                .value(formatDate(Instant.fromEpochMilliseconds(purse.purseExpiryDate * 1000L), DateFormatStyle.LONG))
            purseUiBuilder.item().title("Autoload Amount").value(purse.autoLoadAmount)
            purseUiBuilder.item().title("CAN").value(purse.can)
            purseUiBuilder.item().title("CSN").value(purse.csn)

            val transactionUiBuilder = cardUiBuilder.item().title("Last Transaction Information")
            transactionUiBuilder.item().title("TRP").value(purse.lastTransactionTRP)
            transactionUiBuilder.item().title("Credit TRP").value(purse.lastCreditTransactionTRP)
            transactionUiBuilder.item().title("Credit Header").value(purse.lastCreditTransactionHeader)
            transactionUiBuilder.item().title("Debit Options").value(purse.lastTransactionDebitOptionsByte)

            val otherUiBuilder = cardUiBuilder.item().title("Other Purse Information")
            otherUiBuilder.item().title("Logfile Record Count").value(purse.logfileRecordCount)
            otherUiBuilder.item().title("Issuer Data Length").value(purse.issuerDataLength)
            otherUiBuilder.item().title("Issuer-specific Data").value(purse.issuerSpecificData)
        }

        return cardUiBuilder.build()
    }

    companion object {
        fun create(
            tagId: ByteArray,
            scannedAt: Instant,
            purses: List<CEPASPurse>,
            histories: List<CEPASHistory>
        ): CEPASCard {
            return CEPASCard(tagId, scannedAt, purses, histories)
        }
    }
}
